Bottom: 7f01430c67fa9f0fb8e18ef89eab5b7bb0a1c033
Top:    4d4127992f825a9e3dd7cd1106d1b403a84433ac
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-09-05 00:36:15 +0100

Only assign stencil planes where sub-compositing is required

That is, only for visible layers, and not the far-side, mask or
rat layers.


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index b4e45f9..72fb7e9 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -221,41 +221,59 @@ int
 ghid_set_layer (const char *name, int group, int empty)
 {
   render_priv *priv = gport->render_priv;
-
-  end_subcomposite ();
-  start_subcomposite ();
-
-  /* Drawing is already flushed by {start,end}_subcomposite */
-  hidgl_set_depth (compute_depth (group));
+  bool group_visible = false;
+  bool subcomposite = true;
 
   if (group >= 0 && group < max_group)
     {
       priv->trans_lines = true;
-      return is_layer_group_visible (group);
+      subcomposite = true;
+      group_visible = is_layer_group_visible (group);
     }
   else
     {
       switch (SL_TYPE (group))
 	{
 	case SL_INVISIBLE:
-	  return PCB->InvisibleObjectsOn;
+	  priv->trans_lines = false;
+	  subcomposite = false;
+	  group_visible = PCB->InvisibleObjectsOn;
+	  break;
 	case SL_MASK:
-	  return TEST_FLAG (SHOWMASKFLAG, PCB);
+	  priv->trans_lines = true;
+	  subcomposite = false;
+	  group_visible = TEST_FLAG (SHOWMASKFLAG, PCB);
+	  break;
 	case SL_SILK:
 	  priv->trans_lines = true;
-	  return PCB->ElementOn;
+	  subcomposite = true;
+	  group_visible = PCB->ElementOn;
+	  break;
 	case SL_ASSY:
-	  return 0;
+	  break;
 	case SL_PDRILL:
 	case SL_UDRILL:
-	  return 1;
+	  priv->trans_lines = true;
+	  subcomposite = true;
+	  group_visible = true;
+	  break;
 	case SL_RATS:
-	  if (PCB->RatOn)
-	    priv->trans_lines = true;
-	  return PCB->RatOn;
+	  priv->trans_lines = true;
+	  subcomposite = false;
+	  group_visible = PCB->RatOn;
+	  break;
 	}
     }
-  return 0;
+
+  end_subcomposite ();
+
+  if (group_visible && subcomposite)
+    start_subcomposite ();
+
+  /* Drawing is already flushed by {start,end}_subcomposite */
+  hidgl_set_depth (compute_depth (group));
+
+  return group_visible;
 }
 
 static void
@@ -1872,7 +1890,7 @@ ghid_draw_everything (BoxTypePtr drawn_area)
 
   /* Draw pins, pads, vias below silk */
   if (global_view_2d) {
-    gui->set_layer (NULL, SWAP_IDENT ? solder_group : component_group, 0);
+    start_subcomposite ();
 
     if (!TEST_FLAG (THINDRAWFLAG, PCB)) {
       /* Mask out drilled holes */
@@ -1889,7 +1907,7 @@ ghid_draw_everything (BoxTypePtr drawn_area)
     if (PCB->PinOn) r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_callback, NULL);
     if (PCB->ViaOn) r_search (PCB->Data->via_tree, drawn_area, NULL, via_callback, NULL);
 
-    gui->end_layer ();
+    end_subcomposite ();
   }
 
   /* Draw the solder mask if turned on */
